src/fetch_package_test.js

```diff
@@ -105,3 +107,94 @@ describe("fetchPackage", () => {
         mockScope.done();
     });
 });
+
+describe("fetchPackage with cache", () => {
+    let mockScope;
+
+    before(() => {
+        nock.disableNetConnect();
+        nock.enableNetConnect("127.0.0.1");
+    });
+
+    beforeEach(() => {
+        global._fetched = undefined;
+        mockScope = nock("https://www.ka.org");
+        sinon.stub(args, "useCache").get(() => true);
+    });
+
+    afterEach(() => {
+        flushCache();
+        global._fetched = undefined;
+        nock.cleanAll();
+        sinon.restore();
+    });
+
+    it("should only fetch once for multiple requests", async () => {
+        // Arrange
+        mockScope.get("/ok.js").reply(200, "global._fetched = 'yay!';");
+        mockScope.get("/ok.js").reply(200, "global._fetched = 'ignored';");
+
+        // Act
+        const result0 = await fetchPackage("https://www.ka.org/ok.js", "TEST");
+        const result1 = await fetchPackage("https://www.ka.org/ok.js", "TEST");
+
+        // Assert
+        assert.equal(result0.content, result1.content);
+        // We should still have pending mocks; the second request
+        // should never have gotten sent.
+        assert.notEqual(0, mockScope.pendingMocks().length);
+    });
+
+    it("should retry on 4xx even with cache", async () => {
+        // Arrange
+        mockScope.get("/ok.js").reply(404, "global._fetched = 'boo';");
+        mockScope.get("/ok.js").reply(404, "global._fetched = 'boo';");
+        mockScope.get("/ok.js").reply(404, "global._fetched = 'boo';");
+
+        // Act
+        try {
+            await fetchPackage("https://www.ka.org/ok.js", "TEST");
+        } catch (e) {
+            // Assert
+            assert.equal(404, e.response.status);
+            assert.equal(0, mockScope.pendingMocks().length);
+            mockScope.done();
+            return;
+        }
+        throw new Error("Should have failed on 4xx");
```

Comment:
Seeing it again, it's a clumsy flow having to put the assert in a catch like this. It will be nice if we get chance to move to jest and make this a little nicer with `await expect().rejects....`